openmc/data/resonance_covariance.py

```diff
@@ -0,0 +1,841 @@
+    
+        Parameters
+        ----------
+        n_samples : int
+            The number of samples to produce
+        use_subset : bool, optional
+            Flag on whether to sample from an already produced subset
+    
+        Returns
+        -------
+        samples : list of openmc.data.ResonanceCovarianceRange objects 
+            List of samples size [n_samples]
+    
+        """
+        if use_subset==False:
+            parameters = self.parameters
+            cov = self.covariance
+        else:
+            if self.parameters_subset is None:
+                raise ValueError('No subset of resonances defined')
+            parameters = self.parameters_subset
+            cov = self.cov_subset
+
+        nparams,params = parameters.shape
+        cov = cov + cov.T - np.diag(cov.diagonal()) #symmetrizing covariance matrix
+        covsize = cov.shape[0]
+        formalism = self.formalism
+        mpar = self.mpar
+        samples = []
+    
+    
+        # Handling MLBW sampling
+        if formalism == 'mlbw' or formalism == 'slbw':
+            if mpar == 3:
+                param_list = ['energy','neutronWidth','captureWidth']
+                mean_array = pd.DataFrame.as_matrix(parameters[param_list])
+                spin = pd.DataFrame.as_matrix(parameters['J'])
+                l_value = pd.DataFrame.as_matrix(parameters['L'])
+                gf = pd.DataFrame.as_matrix(parameters['fissionWidth'])
+                gx = pd.DataFrame.as_matrix(parameters['competitiveWidth'])
+                mean = mean_array.flatten()
+                for i in range(n_samples):
+                    sample = np.random.multivariate_normal(mean,cov)
+                    energy = sample[0::3]
+                    gn = sample[1::3]
+                    gg = sample[2::3]
+                    gt = gn + gg + gf
+                    records = []
+                    for j, E in enumerate(energy):
+                        records.append([energy[j], l_value[j], spin[j], gt[j], gn[j],
+                                        gg[j], gf[j], gx[j]])
+                    columns = ['energy', 'L', 'J', 'totalWidth', 'neutronWidth',
+                           'captureWidth', 'fissionWidth', 'competitiveWidth']
+                    sample_params = pd.DataFrame.from_records(records, columns=columns)
+                    samples.append(sample_params)
+    
+            elif mpar == 4:
+                param_list = ['energy','neutronWidth','captureWidth','fissionWidth']
+                mean_array = pd.DataFrame.as_matrix(parameters[param_list])
+                spin = pd.DataFrame.as_matrix(parameters['J'])
+                l_value = pd.DataFrame.as_matrix(parameters['L'])
+                gx = pd.DataFrame.as_matrix(parameters['competitiveWidth'])
+                mean = mean_array.flatten()
+                for i in range(n_samples):
+                    sample = np.random.multivariate_normal(mean,cov)
```

Comment:
One more thing to consider -- it looks like the `multivariate_normal` function also takes an optional `size` argument that would allow you to generate multiple samples with a single function call. This should help improve performance when many samples are needed.